pkg/kine/drivers/generic/generic.go

```diff
@@ -381,6 +395,142 @@ func (d *Generic) queryRowPrepared(ctx context.Context, txName, sql string, prep
+		return
+	}
+
+	if d.BatchingMaxQueries < 5 {
+		logrus.WithFields(logrus.Fields{"interval": d.BatchingInterval, "queue_size": d.BatchingMaxQueries}).Print("Queue size is less than 5, disabling batching of insert queries")
+		return
+	}
+
+	logrus.Warn("Batching of write queries is currently experimental and MUST NOT be used in production. Expect data loss.")
+	logrus.WithFields(logrus.Fields{"interval": d.BatchingInterval, "queue_size": d.BatchingMaxQueries}).Print("Enabling batching of insert queries")
+	d.flushCh = make(chan struct{})
+	d.addToBatchQueueCh = make(chan BatchedInsert)
+	d.removeFromBatchQueueCh = make(chan int)
+	d.batchingQueue = make([]BatchedInsert, 0, d.BatchingMaxQueries)
+	d.batchingCond = sync.Cond{L: &d.batchingMu}
+
+	go d.watchBatchQueue(ctx)
+	go d.batchQueueWatcher(ctx)
+
+}
+
+func (d *Generic) batchQueueWatcher(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case item := <-d.addToBatchQueueCh:
+			d.batchingCond.L.Lock()
+			if len(d.batchingQueue) >= d.BatchingMaxQueries {
+				d.flushCh <- struct{}{}
+				d.batchingCond.Wait()
+			}
+			d.batchingQueue = append(d.batchingQueue, item)
+			d.batchingCond.L.Unlock()
+		}
+	}
+}
+
+func (d *Generic) processBatchedInserts(ctx context.Context, queue []BatchedInsert) ([]int64, error) {
+	valueStrings := make([]string, 0, len(queue))
+	// len(queue) * argument_count per insert
+	valueArgs := make([]interface{}, 0, len(queue)*8)
+
+	for _, insertItem := range queue {
+		valueStrings = append(valueStrings, "(?, ?, ?, ?, ?, ?, ?, ?)")
+		valueArgs = append(valueArgs, insertItem.args...)
+	}
+
+	// Maybe move this into the Generic struct like other queries
+	// Although since we are dynamically generating this we'd only store the template string there
+	stmt := fmt.Sprintf("INSERT INTO kine(name, created, deleted, create_revision, prev_revision, lease, value, old_value) VALUES%s", strings.Join(valueStrings, ","))
+
+	prepared, err := d.DB.Prepare(stmt)
+	if err != nil {
+		logrus.WithError(err).Error("Process batch queue error")
+		return nil, err
+	}
+
+	row, err := d.executePrepared(ctx, "batch_insert_last_insert_id_sql", stmt, prepared, valueArgs...)
+
+	if err != nil {
+		logrus.WithError(err).Error("Process batch queue error")
+		return nil, err
+	}
+
+	lastId, err := row.LastInsertId()
+
+	if err != nil {
+		logrus.WithError(err).Error("Process batch queue error")
+		return nil, err
+	}
+
+	var ids []int64
+
+	// Ideally instead of counting back from LastInsertId, sqlite supports RETURNING
+	// so something like "INSERT INTO .... RETURNING" id would be better.
+	// However db.Prepare complains about wrong syntax when RETURNING is used
+	for i := len(queue); i > 0; i-- {
+		ids = append(ids, lastId-int64(i)+1)
+	}
+
+	return ids, nil
+}
+
+func (d *Generic) processBatchQueue(ctx context.Context) {
```

Comment:
When executing the processBatchQueue we lock the queue, create a copy of the items we want to processBatchedInserts then we unlock the queue and we do the processing. At the end of the processing we lock the queue and we remove the section that we processed. This processBatchQueue is called (from watchBatchQueue) either if we flush or if the BatchingInterval expires. It should be possible to have multiple calls processing the same items from the queue. While the first call is in processBatchedInserts a second can enter the processBatchQueue and take the same queue items to process.